### PR TITLE
ChangeCoreInstallFolder

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

Change DynamoCore Install location to use major version of DynamoCore only. A different PR is sent to DynamoRevitInstaller repo to make similar changes to DynamoRevit install path.

image

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@ramramps 

### FYIs

@smangarole @jnealb @kronz @Racel

### CODE REVIEW

File: src/DynamoInstall/Product.wxs [hunk 1]
```diff
@@ -79,7 +79,7 @@
     <DirectoryRef Id="TARGETDIR">
       <Component Id="ProductRegistry" Win64="yes">
         <RegistryKey Root="HKLM"
-                     Key="SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\$(var.ProductName) $(var.Major).$(var.Minor)"
+                     Key="SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\$(var.ProductName) $(var.Major)"
```
Comment (QilongTang) [inline src/DynamoInstall/Product.wxs]: This make sure when uninstalling, uninstaller looks for major version folder only as well, the uninstall logic is not touched, so side by side install/uninstall is still valid

**Comment by jnealb** [comment]
@QilongTang @smangarole @mjkkirschner @Racel @kronz What task is addressed by this change? We need to be very careful about Install changes given the ever-widening matrix we have to support.

**Comment by QilongTang** [comment]
@jnealb https://jira.autodesk.com/browse/QNTM-1359, yes we are being careful. This change will start from Dynamo 2.0 and not affecting any existing 1.X folder structure, Dynamo 2.1 will isntall to the same folder and overwrite all the contents inside the folder

**Comment by QilongTang** [comment]
Ping @ramramps

**Comment by ramramps** [comment]
LGTM

**Comment by mjkkirschner** [comment]
🙀

**Comment by QilongTang** [comment]
Found twp changes in this PR is not need which can result in unexpected result, going to open a new PR to revert those two lines

**Comment by jnealb** [comment]
@QilongTang Is there any impact from this change to QNTM-2617 - "Un-install of D4R 2.0 disables the Dynamo Player icon from Revit's Manage tab"?

**Comment by QilongTang** [comment]
@jnealb Not that I can think of, player team manages the dll to control if Revit should enable or disable player button.

**Comment by radumg** [comment]
Hi all, might be late to the party but just wanted to highlight **as a user** that this breaks our existing enterprise-wide rollout of Dynamo packages.
We've found packages are often closely tied to specific versions of Dynamo, working in `1.2` but not `1.3` for example and have designed our rollout mechanisms accordingly. This change would make us unable to roll out different versions of packages for different Dynamo versions post-2.0.

Why would we have a mix of `2.0` and `2.1` (or others) on our environment ? Projects routinely get locked to certain versions of software by the project BIM execution plans, be it Revit, Dynamo, so the `update to latest` rationale can't be applied. We still need to support these users whilst the project is active, which could be years : we still have Revit 2014 + Dynamo 0.8.2 on our system for 3 projects as a concrete example. 

Hopefully this explains our need to be able to deploy different packages/configs to different versions of Dynamo. Curious what this actually fixes/improves.

**Comment by mjkkirschner** [comment]
@kronz @QilongTang ☝️ 
also - it's worth noting that if we want to stick strictly to semantic versioning that not having side by side 2.x makes that a riskier proposition for users and integrators.

**Comment by QilongTang** [comment]
@kronz @mjkkirschner @radumg Can we weight the loss against the gain? It seems the gain here is that we will have a clearer file structure which aligns with how Revit installs and is easier for user to understand and for Revit Patch installation. But the loss reported would be that packages would need constant care since it will be more difficult for people to roll back to a version with package working. However, since DynamoCore is sticking to semantic versioning, I actually was not sure why packages should ever enter into that inconsistent working state starting from 2.0. But maybe I am missing something here.

**Comment by radumg** [comment]
Thanks @QilongTang for the clarifications, here's my notes on the above points :

- `which aligns with how Revit installs` - assumes that Revit is a good model for how things should be deployed, nowhere near in my experience, changes every year and it's always a mess. I dread the time of year when I have to update our Revit deployment and bundling Dynamo in 2017 didn't help (on the contrary!)
- ` is easier for user to understand` - only at first glance as it reduces the number of folders, however it obscures actual versioning details. In our case, our general users have no clue, don't care or can't access (permissions) where Dynamo is installed, only people who care about that are the ones deploying it (i.e. me) and the ones supporting it (Design Technology team).
- `semantic versioning, not sure why packages should ever enter into inconsistent working state` 1.1 -> 1.3 was not a major/breaking change in semantic versioning terms but the changes were substantial enough that it required many packages to be updated to work correctly again. I anticipate similar scenarios for 2.1, 2.2 (if they ever happen) - we currently have a backlog of hundreds of scripts that need updating due to the 1.2/1.3 changes btw, they likely never will be.

Biggest loss for me is still that we need to be able to have multiple versions installed at the same time and not just limited to major 1.x & 2.x releases - and be able to support them with different package versions rolled out. I have 0.9, 1.2, 1.3.1 and 2.0 installed on my machine to cater to the different projects we support, as do many of our users. The previous example of 3 projects translates to roughly 30 users and that's only about one Revit version in one office alone.

**Comment by kronz** [comment]
@radumg So, to clarify . . . you are saying that you have side by side installations of 1.0-1.3?  Not sure how you would do that.  Or do I misunderstand?  The issue we are trying to address is that we have seen implementors referencing, say, a 1.1 install folder. Then their software stops working when 1.2 removes 1.1, not because of an incompatibility, but simply because they hardwired the path to that version.  Structuring the folders as 1.1, 1.2, etc, was originally an error, and it incorrectly implies that point releases can coexist.

**Comment by Racel** [comment]
As per our discussion with @radumg, the new behavior is acceptable and should not break existing workflows for package management. @radumg - please let us know if any issues arise when you test out the daily builds.

**Pull request merged**